### PR TITLE
Discriminate generics inside lists

### LINKED ISSUE #3432: buggy generic interfaces
```
import strawberry
from typing import Optional, Generic, TypeVar, List
from strawberry.scalars import JSON

GenericType = TypeVar("GenericType")

@strawberry.interface
class BlockInterface:
    id: strawberry.ID
    disclaimer: Optional[str] = strawberry.field(default=None, description='Richtext')

@strawberry.type
class JsonBlock(BlockInterface):
    data: JSON

@strawberry.type
class BlockRowType(BlockInterface, Generic[GenericType]):
    total: int
    items: List[GenericType]

@strawberry.type
class Query:
    
    @strawberry.field
    def blocks(self) -> List[BlockInterface]:
        return [
            BlockRowType(id="3", total=3, items=["a", "b", "c"]),
            BlockRowType(id="1", total=1, items=[1, 2, 3, 4]),
            JsonBlock(id="2", data={"a": 1}),
        ]

schema = strawberry.Schema(query=Query, types=[BlockRowType[int], JsonBlock, BlockRowType[str]])
```

throws: `Int cannot represent non-integer value: 'a'`

See this gists for more examples:

https://play.strawberry.rocks/?gist=e236f23fee40493f1952ad88cb05df80
https://play.strawberry.rocks/?gist=f56308e2f925eb05c37d82b4ad70efd8
https://play.strawberry.rocks/?gist=3c48b2f9e848042e87a9f9c9da143994

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/strawberry-graphql) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/strawberry-graphql/strawberry/issues/3432">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/strawberry-graphql/strawberry/issues/3432/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by patrick91**
Still WIP, should fix #3432, but right now it only fixes
the union part 😊

### CODE REVIEW

**Comment by botberry** [comment]
## Apollo Federation Subgraph Compatibility Results


<table>
	<thead>
		<tr>
			<th>Federation 1 Support</th>
			<th>Federation 2 Support</th>
		</tr>
	</thead>
	<tbody>
    <tr>
      <td>
        <table>
          <tr><th><code>_service</code></th><td>🟢</td></tr>
          <tr><th><code>@key (single)</code></th><td>🟢</td></tr>
          <tr><th><code>@key (multi)</code></th><td>🟢</td></tr>
          <tr><th><code>@key (composite)</code></th><td>🟢</td></tr>
          <tr><th><code>repeatable @key</code></th><td>🟢</td></tr>
          <tr><th><code>@requires</code></th><td>🟢</td></tr>
          <tr><th><code>@provides</code></th><td>🟢</td></tr>
          <tr><th><code>federated tracing</code></th><td>🔲</td></tr>
        </table>
      </td>
      <td>
        <table>
          <tr><th><code>@link</code></th><td>🟢</td></tr>
          <tr><th><code>@shareable</code></th><td>🟢</td></tr>
          <tr><th><code>@tag</code></th><td>🟢</td></tr>
          <tr><th><code>@override</code></th><td>🟢</td></tr>
          <tr><th><code>@inaccessible</code></th><td>🟢</td></tr>
          <tr><th><code>@composeDirective</code></th><td>🟢</td></tr>
          <tr><th><code>@interfaceObject</code></th><td>🟢</td></tr>
        </table>
      </td>
		</tr>
	</tbody>
</table>


Learn more:
* [Apollo Federation Subgraph Specification](https://www.apollographql.com/docs/federation/subgraph-spec/)
* [Compatibility Tests](https://github.com/apollographql/apollo-federation-subgraph-compatibility/blob/main/COMPATIBILITY.md)

**Comment by patrick91** [comment]
Random thought we might want to add support for something like this:

```python
return BlockRowType[str](total=3, items=[]),
```

because when we have 0 items when don't know which type we should use, so for now we fallback to the first

File: strawberry/types/types.py [hunk 1]
```diff
@@ -170,20 +171,48 @@ def is_implemented_by(self, root: Type[WithStrawberryObjectDefinition]) -> bool:
 
         # Check the mapping of all fields' TypeVars
         for generic_field in type_definition.fields:
```
Comment (patrick91) [inline strawberry/types/types.py:173]: this should be maybe_generic_field? :D

**Comment by botberry** [comment]
# Pre-release
:wave:

Releasing commit [cddafaea921ef41ca03207080679ed9487712cc5] to PyPi as pre-release! :package:

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.228.0.dev.1713643365** [cddafaea921ef41ca03207080679ed9487712cc5] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.228.0.dev.1713643365
```

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release improves our support for generic types, now using the same the same
generic multiple times with a list inside an interface or union is supported,
for example the following will work:

```python
import strawberry


@strawberry.type
class BlockRowType[T]:
    items: list[T]


@strawberry.type
class Query:
    @strawberry.field
    def blocks(self) -> list[BlockRowType[str] | BlockRowType[int]]:
        return [
            BlockRowType(id=strawberry.ID("3"), items=["a", "b", "c"]),
            BlockRowType(id=strawberry.ID("1"), items=[1, 2, 3, 4]),
        ]


schema = strawberry.Schema(query=Query)
```



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```

File: strawberry/types/types.py [hunk 2]
```diff
@@ -169,21 +170,43 @@ def is_implemented_by(self, root: Type[WithStrawberryObjectDefinition]) -> bool:
...
+            if not field.is_graphql_generic:
+                continue
+
+            value = getattr(root, field.name)
+            generic_field_type = field.type
+
+            while isinstance(generic_field_type, StrawberryList):
+                generic_field_type = generic_field_type.of_type
+
+                assert isinstance(value, (list, tuple))
```
[the remark this answers is not in the document]
Comment (patrick91) [reply strawberry/types/types.py:183]: I think I'll keep the check simple for now, we might want to check sequence though, but I'd do that after a bug report 😊

File: strawberry/schema/schema_converter.py [hunk 3]
```diff
@@ -450,7 +451,27 @@ def resolve_type(
...
+
+                    # here we don't all the implementations of the generic
+                    # we need to find a way to find them, for now maybe
+                    # we can follow the union's approach and iterate over
+                    # all the types in the schema, but we should probably
+                    # optimize this
+
+                    return_type: Optional[GraphQLType] = None
+
+                    for possible_concrete_type in self.type_map.values():
```
[the remark this answers is not in the document]
Comment (patrick91) [reply strawberry/schema/schema_converter.py:464]: yup, there's a comment about this, a future improvement for sure 😊

https://github.com/strawberry-graphql/strawberry/issues/3465

File: strawberry/types/types.py [hunk 4]
```diff
@@ -169,21 +170,43 @@
...
+                if len(value) == 0:
+                    # We can't infer the type of an empty list, so we just
+                    # return the first one we find
+                    return True
+
+                value = value[0]
+
+            if isinstance(generic_field_type, StrawberryTypeVar):
+                type_var = generic_field_type.type_var
+            # TODO: I don't think we support nested types properly
```
[the remark this answers is not in the document]
Comment (patrick91) [reply strawberry/types/types.py:194]: https://github.com/strawberry-graphql/strawberry/issues/3466

**Comment by patrick91** [comment]
@Speedy1991 let me know if you have time to try this!

**Comment by Speedy1991** [comment]
I'm not 100% sure that my code is correct, but this example fails

https://play.strawberry.rocks/?gist=d7d4da4f8b8fe245e645736ba5708b83

**Comment by Speedy1991** [comment]
> Random thought we might want to add support for something like this:
> 
> ```python
> return BlockRowType[str](total=3, items=[]),
> ```

Could this lead to problems on the client site? E.g. Apollo uses cache normalization (not sure about relay - didn't used it for years).  I _guess_ this should be ok as long as the return type matches the possible types for this field - just to make sure :)

**Pull request merged**